Several randomness improvements

sipa opened the pull request:
This PR contains a number of vaguely-related improvements to the random module.

The specific changes and more detailed rationale is in the commit messages, but the highlights are:

* `XoRoShiRo128PlusPlus` (previously a test-only RNG) moves to random.h and becomes `InsecureRandomContext`, which is even faster than `FastRandomContext` but non-cryptographic. It also gets all helper randomness functions (`randrange`, `fillrand`, ...), making it a lot more succinct to use.
* During tests, **all** randomness is made deterministic (except for `GetStrongRandBytes`) but non-repeating (like `GetRand()` used to be when `g_mock_deterministic_tests` was used), either fixed, or from a random seed (overridden by env var).
* Several infrequently used top-level functions (`GetRandMillis`, `GetRandMicros`, `GetExponentialRand`) are converted into member functions of `FastRandomContext` (and `InsecureRandomContext`).
* `GetRand<T>()` (without argument) can now return the maximum value of the type (previously e.g. `GetRand<uint32_t>()` would never return 0xffffffff).

sipa committed:
random: refactor: move rand* utilities to RandomMixin

Rather than make all the useful types of randomness be exclusive to
FastRandomContext, move it to a separate RandomMixin class where it can be reused by
other RNGs.

A Curiously Recurring Template Pattern (CRTP) is used for this, to provide the ability
for individual RNG classes to override one or more randomness functions, without
needing the runtime-cost of virtual classes.

Specifically, RNGs are expected to only provide fillrand and rand64, while all others
are derived from those:
- randbits
- randrange
- randbytes
- rand32
- rand256
- randbool
- rand_uniform_delay
- rand_uniform_duration
- min(), max(), operator()(), to comply with C++ URBG concept.

sipa committed:
random: Improve RandomMixin::randbits

The previous randbits code would, when requesting more randomness than available
in its random bits buffer, discard the remaining entropy and generate new.

Benchmarks show that it's usually better to first consume the existing randomness
and only then generate new ones. This adds some complexity to randbits, but it
doesn't weigh up against the reduced need to generate more randomness.

sipa committed:
random: add RandomMixin::randbits with compile-known bits

In many cases, it is known at compile time how many bits are requested from
randbits. Provide a variant of randbits that accepts this number as a template,
to make sure the compiler can make use of this knowledge. This is used immediately
in rand32() and randbool(), and a few further call sites.

sipa committed:
random: modernize XoRoShiRo128PlusPlus a bit

Make use of C++20 functions in XoRoShiRo128PlusPlus.

sipa committed:
random: move XoRoShiRo128PlusPlus into random module

This is preparation for making it more generally accessible.

sipa committed:
random: convert XoRoShiRo128PlusPlus into full RNG

Convert XoRoShiRo128PlusPlus into a full RandomMixin-based RNG class,
providing all utility functionality that FastRandomContext has. In doing so,
it is renamed to InsecureRandomContext, highlighting its non-cryptographic
nature.

To do this, a fillrand fallback is added to RandomMixin (where it is used by
InsecureRandomContext), but FastRandomContext still uses its own fillrand.

sipa committed:
tests: overhaul deterministic test randomness

The existing code provides two randomness mechanisms for test purposes:
- g_insecure_rand_ctx (with its wrappers InsecureRand*), which during tests is
  initialized using either zeros (SeedRand::ZEROS), or using environment-provided
  randomness (SeedRand::SEED).
- g_mock_deterministic_tests, which controls some (but not all) of the normal
  randomness output if set, but then makes it extremely predictable (identical
  output repeatedly).

Replace this with a single mechanism, which retains the SeedRand modes to control
all randomness. There is a new internal deterministic PRNG inside the random
module, which is used in GetRandBytes() when in test mode, and which is also used
to initialize g_insecure_rand_ctx. This means that during tests, all random numbers
are made deterministic. There is one exception, GetStrongRandBytes(), which even
in test mode still uses the normal PRNG state.

This probably opens the door to removing a lot of the ad-hoc "deterministic" mode
functions littered through the codebase (by simply running relevant tests in
SeedRand::ZEROS mode), but this isn't done yet.

sipa committed:
random: make GetRand() support entire range (incl. max)

The existing code uses GetRand(nMax), with a default value for nMax, where nMax is the
range of values (not the maximum!) that the output is allowed to take. This will always
miss the last possible value (e.g. GetRand<uint32_t>() will never return 0xffffffff).

Fix this, by moving the functionality largely in RandomMixin, and also adding a
separate RandomMixin::rand function, which returns a value in the entire (non-negative)
range of an integer.

sipa committed:
random: convert GetRand{Micros,Millis} into randrange

There are only a few call sites of these throughout the codebase, so
move the functionality into FastRandomContext, and rewrite all call sites.

This requires the callers to explicit construct FastRandomContext objects,
which do add to the verbosity, but also make potentially apparent locations
where the code can be improved by reusing a FastRandomContext object (see
further commit).

sipa committed:
random: convert GetExponentialRand into rand_exp_duration

sipa committed:
random: cleanup order, comments, static

dergoegge commented:
Concept ACK

There are a few places in the fuzz tests where this will allow to easily replace `FastRandomContext` with a `InsecureRandomContext`, which is beneficial for performance (e.g. the addrman harnesses partially fill the addrman with addresses from rng) and we don't need cryptographic rng there anyway.

> During tests, all randomness is made deterministic 

Great, this should help with #29018

dergoegge commented:
> What's the impact on the fuzz corpus of switching to a different (?) deterministic RNG?

I would expect that switching to a **different** rng should have no meaningful effect on the corpus itself. The corpus for a particular harness might change but the coverage for the code we intend to test should remain the same. This is because using rng in a fuzz harness only makes sense in very rare cases. It should never be used in a way that can significantly affect the coverage reached, otherwise there is no point in using a coverage-guided fuzzer, we could just pipe /dev/random to our harnesses.

For example, if we need to populate some data that we don't really expect to have an impact on the thing we are testing, we might use rng instead of consuming from the fuzz input (we do this in the p2p transport harnesses to fill message contents, which are essentially irrelevant to the transport logic).

Switching to **deterministic** rng can cause a corpus' coverage to grow because coverage-guided feedback loops start working more reliably when the code under test is deterministic. This can vary from harness to harness, but we've seen coverage-guided fuzzers find bugs once we've improved on determinism.

sipa committed:
tests: make fuzz tests (mostly) deterministic with fixed seed

EthanHeilman commented:
@sipa I plan to do a review of this next week

dergoegge commented on src/net_processing.cpp:
```diff
@@ -5501,7 +5503,7 @@ void PeerManagerImpl::MaybeSendFeefilter(CNode& pto, Peer& peer, std::chrono::mi
             peer.m_fee_filter_sent = filterToSend;
         }
-        peer.m_next_send_feefilter = GetExponentialRand(current_time, AVG_FEEFILTER_BROADCAST_INTERVAL);
+        peer.m_next_send_feefilter = current_time + FastRandomContext().rand_expo_duration(AVG_FEEFILTER_BROADCAST_INTERVAL);
```
https://github.com/bitcoin/bitcoin/pull/28558 made PeerManager own a FastRandomContext, so we could (should?) use `m_rng` here instead (otherwise `PeerManager::Options::deterministic_rng` still only applies to some of the randomness).

Since this PR kind of makes individual "make this component deterministic" options redudant, we could consider reverting #28558 (not necessarily in this PR)?

I was thinking that in the long run we could break the dependencies between components and the specific rng they use (maybe something like `template<RandomNumberGenerator R> class PeerManager { ... }`), which would allow more fine grained mocking than a global "make rng deterministic" in tests (e.g. we could have a "rng" type that consumes from a `FuzzedDataProvider`). I guess this can be done by using globals as well.

dergoegge reviewed (approved):
Code review ACK 636e4864516782b7e6e33be2b2ec7743afaae8ba

theuni reviewed (commented):
Concept ACK and first-pass code review ACK excluding tests.

I don't feel qualified to review the rand_expo_duration changes.

I manually verified the randbits impls with some local tests as the shifts weren't intuitive to me at first glance.

theuni commented on src/random.h:
```diff
@@ -351,19 +363,19 @@ class FastRandomContext : public RandomMixin<FastRandomContext>
     }
 
-    /** Fill a byte Span with random bytes. */
+    /** Fill a byte Span with random bytes. This overrides the RandomMixin version. */
```
This is kinda sneaky and surprising (to me at least) to see with CRTP.

It would be nice to specify as part of the concept which functions are _not_ allowed to be overridden, but sadly 
[std::experimental::detected_t](https://en.cppreference.com/w/cpp/experimental/is_detected) is.. well.. experimental.

theuni commented on src/random.h:
```diff
@@ -135,50 +135,68 @@ void RandAddPeriodic() noexcept;
 
-    /** Generate a random 64-bit integer. */
-    uint64_t rand64() noexcept
+    RandomMixin& operator=(RandomMixin&& other) noexcept
```
While you're at it, how about replacing these move assignments with `Reseed()` functions?

I find the idea of moving an rng to be unintuitive. And as far as I can tell, every case of 

`RandomMixin = foo` and `FastRandomContext = foo` and `InsecureRandomContext = foo` (all in tests) really just means "reseed".

theuni commented on src/random.h:
```diff
@@ -215,6 +215,7 @@ class RandomMixin
             // number becomes the new bitbuf.
             uint64_t gen = Impl().rand64();
             ret = (gen << bitbuf_size) | bitbuf;
+            assert(bits >= bitbuf_size);
```
I tried to track down if/how it could be a race condition, but given that it's in init and all locks seems to be in place, I don't see how that could possibly be the case.

sipa committed:
random: replace construct/assign with explicit Reseed()

achow101 reviewed (commented):
Reviewed up to 64bcc828cde51a5da551b49a31b9fe44e0a255c3

maflcko reviewed (commented):
e7676b6c52672f557f0e2b036d5e877393f1ce46 🎩

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: e7676b6c52672f557f0e2b036d5e877393f1ce46 🎩
cxzra7fTi/Rsr6+8RvbXnVktwr1MJ/eJJsVn5hm1X2uUnnz3rNYpzIbtPyGWi9dcTAB7fkupq9pQeUC8sQG4AQ==
```

</details>

sipa committed:
random: write rand256() in function of fillrand()

sipa committed:
random: move rand256() and randbytes() to .h file

sipa committed:
random: add a few noexcepts to FastRandomContext

sipa committed:
random: use BasicByte concept in randbytes

sipa committed:
random: use LogError for init failure

sipa committed:
xoroshiro128plusplus: drop comment about nonexisting copy()

EthanHeilman commented on src/random.cpp:
```diff
@@ -479,6 +438,13 @@ class RNGState {
             // Store the last 32 bytes of the hash output as new RNG state.
             memcpy(m_state, buf + 32, 32);
+            // Handle requests for deterministic randomness.
+            if (allow_deterministic && m_deterministic_prng.has_value()) {
```
Does `allow_deterministic=false` and `m_deterministic_prng.has_value()=true` imply that something has gone wrong? Should we log an error or take some action or are there cases where this is expected behavior?

sipa committed:
net: use GetRandMicros for cache expiration

This matches the data type of m_cache_entry_expiration.

EthanHeilman commented on src/random.cpp:
```diff
@@ -345,64 +348,6 @@ static void Strengthen(const unsigned char (&seed)[32], SteadyClock::duration du
 #endif
 
-/** Get 32 bytes of system entropy. */
-void GetOSRand(unsigned char *ent32)
```
That makes sense. This improved organization is worth it.

EthanHeilman commented on src/random.cpp:
```diff
@@ -457,11 +406,21 @@ class RNGState {
      * If this function has never been called with strong_seed = true, false is returned.
+     *
+     * If allow_deterministic is true, and MakeDeterministic has been called before, output
+     * from the deterministic PRNG is output instead.
```
1. If I understand, you don't want to replace `MixExtract` entirely when the deterministic test RNG is used, because want `MixExtract` to still run and add/mix entropy in case someone decides to generate a keying material while the deterministic test RNG is on. This was what I was missing.

2. I like the name deterministic mode because it makes it more clear this is a RNG wide change.

3. Thought I had after waking up this morning was to change this to `bool fail_on_deterministic_mode`. Thus if someone had the deterministic test RNG activated and then tried to generate a key it would throw an error because `fail_on_deterministic_mode=true`. That said, I'm not sure this works with the current approach since if someone calls SeedStrengthen with the deterministic test RNG activated it would fail and it is reasonable to call SeedStrengthen in tests/fuzzing.

EthanHeilman reviewed (approved):
ACK 125509f395a214465ebf379164e41e4c6d19d443
I have not tested the code. All my comments have been addressed.

theuni commented on src/net_processing.cpp:
```diff
@@ -508,7 +508,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Implement PeerManager */
-    void StartScheduledTasks(CScheduler& scheduler) override;
+    void StartScheduledTasks(CScheduler& scheduler) override EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
```
Lol, I even spent a few hours looking directly at this and trying to figure out if this could be a threading problem.

🤦

I suppose I trusted the annotations so much I didn't even notice that the code didn't take the lock. That's embarrassing!

Looks like [this is our issue](https://github.com/llvm/llvm-project/pull/67520).

I'll give that patched branch a shot next week and see if it turns up anything else in our codebase. Will chime in upstream with a +1 as well (assuming it works).

theuni replied:
I used the WIP above to check for other issues like this in our codebase. It only turned up a few false-positives because of duplicate annotations in declarations + definitions. Thankfully there were no cases of them _only_ in definitions. PR'd here: #30316.

sipa committed:
net, net_processing: use existing RNG objects more

PeerManagerImpl, as well as several net functions, already have existing
FastRandomContext objects. Reuse them instead of constructing new ones.

sipa committed:
random: improve precision of MakeExponentiallyDistributed

achow101 commented:
75bcc1057bbf7d8e0803df09780f4237c10c4264 says "rand_exp_delay" in the commit message, but the actual function is "rand_exp_duration".

maflcko reviewed (approved):
Left some nits/questions. Happy to re-ACK if you address them.

ACK f0e5681e97450b696de26ff21e61f961370a9946 📲

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK f0e5681e97450b696de26ff21e61f961370a9946 📲
AB6nWi2n4+9yfQdFBUUnXA6K+p5+MCJdNp0EwxcHMQLi2gsdpbpVFB1JswtGdQXvPjEmoK3HiR1ROg5SCYouBg==
```

</details>

sipa committed:
random: get rid of GetRand by inlining

maflcko reviewed (approved):
Left some more nits. Feel free to ignore.

achow101 commented:
ACK ce8094246ee95232e9d84f7e37f3c0a43ef587ce

maflcko reviewed (approved):
re-ACK ce8094246ee95232e9d84f7e37f3c0a43ef587ce 🐈

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK ce8094246ee95232e9d84f7e37f3c0a43ef587ce 🐈
iPUv215Td9GqZ8iEiD8qyKJufeqJCMlNIHYw/Ha/B/66jmXmcrePorRaTQ+YqnML9Nkr4A+IaY3dqwDPvXI5Cg==
```

</details>


Reminder for myself: https://github.com/bitcoin/bitcoin/pull/29625#discussion_r1658724899

hodlinator reviewed (approved):
ACK ce8094246ee95232e9d84f7e37f3c0a43ef587ce

Mainly code review. Did not attempt to check thread safety compliance. (Passed `make check` & `test/functional/test_runner.py`).

### e2d1f84858485650ff743753ffa5c679f210a992 - random: make GetRand() support entire range (incl. max)

The commit message title probably should be "random: make GetRand\<T\>() support entire range (incl. max)", since the overloads taking parameters still are exclusive at the end. It felt dangerous that `GetRand<T>(void)` behavior differed in this way from the overloads - happy others suggested the calls be inlined as `randrange()` and `rand<T>()` are much clearer!

dergoegge reviewed (approved):
utACK ce8094246ee95232e9d84f7e37f3c0a43ef587ce

hodlinator commented on src/test/util/net.cpp:
```diff
@@ -126,7 +126,7 @@ std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(int n_candida
             /*m_relay_txs=*/random_context.randbool(),
             /*fBloomFilter=*/random_context.randbool(),
-            /*nKeyedNetGroup=*/random_context.randrange(100),
+            /*nKeyedNetGroup=*/random_context.randrange(100u),
```
Follow-up: #30397

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.
